ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -0,0 +1,279 @@
+    )
+
+    // 3. Verify that the input contracts are active.
+    // FIXME(JM): Does this need to be done separately, or is it enough to
+    // have lookupContract to check these?
+    val activenessCheckResult = true
+
+    // 4. Verify that the submission conforms to the DAML model
+    def lookupContract(coid: AbsoluteContractId) = {
+      val (eid, nid) = absoluteContractIdToLogEntryId(coid)
+      val stateKey = absoluteContractIdToStateKey(coid)
+      for {
+        contractState <- inputState.get(stateKey).flatMap(_.map(_.getContractState))
+        if txLet > parseTimestamp(contractState.getActiveAt) && !contractState.hasActiveAt
+        entry <- inputLogEntries.get(eid)
+        coinst <- lookupContractInstanceFromLogEntry(eid, entry, nid)
+      } yield (coinst)
+    }
+
+    def lookupPackage(pkgId: PackageId) = {
+      val stateKey = DamlStateKey.newBuilder.setPackageId(pkgId.underlyingString).build
+      inputState
+        .get(stateKey)
+        .flatMap(identity)
+        .flatMap { value =>
+          inputLogEntries.get(value.getArchiveEntry)
+        }
+        .flatMap { entry =>
+          entry.getPayloadCase match {
+            case DamlLogEntry.PayloadCase.ARCHIVE =>
+              // NOTE(JM): Engine only looks up packages once, compiles and caches,
+              // provided that the engine instance is persisted.
+              Try(Decode.decodeArchive(entry.getArchive)._2).toOption
+            case _ =>
+              None
+          }
+        }
+    }
+
+    val relTx = Conversions.decodeTransaction(txEntry.getTransaction)
+    val modelCheckResult = engine
+      .validate(relTx, txLet)
+      .consume(lookupContract, lookupPackage, _ => sys.error("contract keys unimplemented"))
+
+    def reject(reason: RejectionReason): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) =
+      (
+        DamlLogEntry.newBuilder
+          .setRecordTime(buildTimestamp(recordTime))
+          .setRejectionEntry(
+            buildRejectionEntry(txEntry, RejectionReason.DuplicateCommand)
+          )
+          .build,
+        Map.empty
+      )
+
+    // FIXME(JM): Refactor this?
+    (dedupCheckResult, letCheckResult, activenessCheckResult, modelCheckResult) match {
+      case (false, _, _, _) =>
+        reject(RejectionReason.DuplicateCommand)
+
+      case (_, false, _, _) =>
+        reject(RejectionReason.MaximumRecordTimeExceeded)
+
+      case (_, _, false, _) =>
+        reject(RejectionReason.Inconsistent)
+
+      case (true, true, true, Left(err)) =>
+        reject(RejectionReason.Disputed(err.msg)) // FIXME(JM): or detailMsg?
+
+      case (true, true, true, Right(())) =>
+        // All checks passed. Return transaction log entry, and update the DAML state
+        // with the committed command and the created and consumed contracts.
+
+        var stateUpdates = scala.collection.mutable.Map.empty[DamlStateKey, DamlStateValue]
```

Comment:
add comments explaining each of the steps. Also helps double-check that all state that is checked up front is modified here.